pkg/csi/manila/controllerserver.go

```diff
@@ -151,11 +151,19 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		}
 	}
 
+	var accessibleTopology []*csi.Topology
+	if cs.d.withTopology {
+		// All requisite topologies are considered valid. Nodes from those zones are required to be able to reach the storage.
+		// The operator is responsible for making sure that provided topology keys are valid and present on the nodes of the cluster.
+		accessibleTopology = req.GetAccessibilityRequirements().GetRequisite()
```

Comment:
Attempt should be made to provision volume accessible from preferred topology right ? not seeing any zone info being passed while creating, Im not sure if applicable for manila though.

Reply:
> not seeing any zone info being passed while creating, Im not sure if applicable for manila though.

It's here:

https://github.com/kubernetes/cloud-provider-openstack/blob/fd6aa3342fb1d917a5fa79e02176075dc9cc5f7d/pkg/csi/manila/volumesource.go#L40-L48

We've added a new storage class parameter `availability` where the operators will specify the Manila AZ.

We're using the storage class parameter AND topology awareness in conjunction because Manila may have its own set of storage AZs, and the k8s cluster may also have its own separate set of compute AZs and that those two sets may or may not overlap, so we've left this as a provisioning detail for the cluster admins: the admin has to know which [`allowedTopologies`](https://kubernetes.io/docs/concepts/storage/storage-classes/#allowed-topologies) are able to  access the Manila AZ specified in the `availability` parameter and define the StorageClass accordingly.

> Attempt should be made to provision volume accessible from preferred topology right ?

Yes you're right, according to the spec we should choose zone(s) from the preferred topologies and fallback to requisite topologies and if none of them are applicable then the provisioning should fail.

The problem is that in case of csi-manila, AFAIK there's no reliable way to check whether a certain storage zone is accessible from a certain compute zone. So I can change this to Preferred topologies, but it won't make much difference, because all the listed topology terms will be used for provisioning anyway.

Reply:
@gman0 @gouthampacha Thanks for the detailed explanation. I understand better now.
I think if delayed binding is not enabled, I agree this preferred/requisite wont make a difference. See my question on above comment , if my understanding correct on delayed binding needed or not.

Reply:
And also , to add my reason why the 'withTopology' flag not needed , even if condition is not there , if Topology not enabled in provisoner, AccessibilityRequirements would be nil. If topology enabled in provisioner , it gets all aggregate topologies , which should be still fine , I think only field will be populated , and there wont be any difference in functionality . Please correct me if Im wrong.

Reply:
@ramineni 

> if Topology not enabled in provisoner, AccessibilityRequirements would be nil. If topology enabled in provisioner , it gets all aggregate topologies , which should be still fine , I think only field will be populated

That's correct!

But so far we've identified a couple of issues with having topology always enabled - all of those seem to be revolving around that Manila may be running outside of Nova environment, or that the (kubernetes) cluster may be consuming Manila storage without actually it being run in OpenStack (i.e. on baremetal or whatever):
* we'll get a silent timeout each time the driver is initializing
* in case we're actually relying on the data from the Metadata Service, we won't be able to fail in case the Metadata svc is (temporarily) unavailable
* if we don't care about topology (i.e. all compute nodes are able to access the Manila-backed storage), we still need to fill `--nodeaz` with something, otherwise we seem to be in conflict with the CSI spec? This could be solved with some placeholder value...although this feels a bit hack-y